tower-http/src/follow_redirect/policy/join.rs

```diff
@@ -0,0 +1,139 @@
+use super::{Action, Attempt, Policy};
+use http::Request;
+
+/// A redirection [`Policy`] that combines the results of two `Policy`s.
+///
+/// See [`join`] for more details.
+#[derive(Clone, Copy, Debug, Default)]
+pub struct Join<A, B> {
+    a: A,
+    b: B,
+}
+
+impl<Bd, E, A, B> Policy<Bd, E> for Join<A, B>
+where
+    A: Policy<Bd, E>,
+    B: Policy<Bd, E>,
+{
+    fn redirect(&mut self, attempt: &Attempt<'_>) -> Action<E> {
+        let a = self.a.redirect(attempt);
+        if a.follows() {
+            self.b.redirect(attempt)
+        } else {
+            a
+        }
+    }
+
+    fn on_request(&mut self, request: &mut Request<Bd>) {
+        self.a.on_request(request);
+        self.b.on_request(request);
+    }
+
+    fn clone_body(&self, body: &Bd) -> Option<Bd> {
+        self.a.clone_body(body).or_else(|| self.b.clone_body(body))
+    }
+}
+
+/// Create a new `Policy` that returns [`Action::follow()`] only if `self` and `other` return
+/// `Action::follow()`.
+///
+/// [`clone_body`][Policy::clone_body] method of the returned `Policy` tries to clone the body
+/// with both policies.
+///
+/// # Example
+///
+/// ```
+/// use bytes::Bytes;
+/// use hyper::Body;
+/// use tower_http::follow_redirect::policy::{self, clone_body_fn, Limited, Policy};
+///
+/// enum MyBody {
+///     Bytes(Bytes),
+///     Hyper(Body),
+/// }
+///
+/// let policy = policy::join::<_, _, _, ()>(
+///     Limited::default(),
+///     clone_body_fn(|body| {
+///         if let MyBody::Bytes(buf) = body {
+///             Some(MyBody::Bytes(buf.clone()))
+///         } else {
+///             None
+///         }
+///     }),
+/// );
+/// ```
+pub fn join<A, B, Bd, E>(a: A, b: B) -> Join<A, B>
+where
+    A: Policy<Bd, E>,
+    B: Policy<Bd, E>,
+{
+    Join { a, b }
+}
```

Comment:
I think this is cool but wondering if it should be method directly on `Policy` and perhaps be called `and`? That way you can do `policy_one.and(policy_two)`. I got confused initially by the name `join` as I was thinking of `futures::join!`.

Reply:
> wondering if it should be method directly on Policy and perhaps be called and?

That was exactly the same design I came up with initially in commit 2fe3fea1e3c2829cbf0474cb2b36194fd289f287, but after I added the error type parameter to `Policy` in commit 04cc691e034342be605242b33f0ff976e8316d2a, the `and` method began to break the type inference:

```
error[E0282]: type annotations needed
  --> src/follow_redirect/policy/mod.rs:90:33
   |
13 | let policy = Limited::default().and(clone_body_fn(|body| {
   |                                 ^^^ cannot infer type for type parameter `E` declared on the trait `Policy`
```

To annotate the type parameter, the caller needs to use the UFCS like `Policy::<_, ()>::and(Limited::default(), clone_body_fn(/* ... */))`, and I thought that a freestanding function would be more ergonomic than UFCS. But I agree that this is still an awkward interface. I will explore other possibilities (maybe an extension trait would work?).

Reply:
I have just added an extension trait `PolicyExt` with `and` and `or` methods and this seems to work fine.

However, I had to implement it for arbitrary`T` without `T: Policy<B, E>` bound to avoid having type parameters in the extension trait, and the methods conflict with methods from other traits named `and` or `or` (I confirmed that they do not conflict with inherent methods like `Option::<_>::and`). I have no idea whether this would cause any issues in practice.